safe_transaction_service/history/services/collectibles_service.py

```diff
@@ -160,55 +164,73 @@ def __init__(self, ethereum_client: EthereumClient, redis: Redis):
-                    raise MetadataRetrievalException(
-                        f"Content-type={content_type} for uri={uri} is not valid, "
-                        f'expected "application/json"'
-                    )
-
-                logger.debug("Got metadata for uri=%s", uri)
+        from ..tasks import chance_to_retrieve_metadata_from_uri
 
-                # Some requests don't provide `Content-Length` on the headers
-                if len(response.content) > self.METADATA_MAX_CONTENT_LENGTH:
-                    raise MetadataRetrievalException(
-                        f"Retrieved content for uri={uri} is too big"
-                    )
+        uri_http = ipfs_to_http(uri)
+        if (
+            not django_cache.get(self.get_redis_metadata_timeout_key(uri))
+            or force_retry
+        ):
+            if not uri_http or not uri_http.startswith("http"):
+                raise MetadataRetrievalException(uri_http)
 
-                return response.json()
-        except (IOError, ValueError) as e:
-            raise MetadataRetrievalException(uri) from e
+            try:
+                logger.debug("Getting metadata for uri=%s", uri_http)
+                with requests.get(uri_http, timeout=15, stream=True) as response:
+                    if not response.ok:
+                        logger.debug("Cannot get metadata for uri=%s", uri_http)
+                        raise MetadataRetrievalException(uri_http)
+
+                    content_length = response.headers.get("content-length", 0)
+                    content_type = response.headers.get("content-type", "")
+                    if int(content_length) > self.METADATA_MAX_CONTENT_LENGTH:
+                        raise MetadataRetrievalException(
+                            f"Content-length={content_length} for uri={uri_http} is too big"
+                        )
+
+                    if "application/json" not in content_type:
+                        raise MetadataRetrievalException(
+                            f"Content-type={content_type} for uri={uri_http} is not valid, "
+                            f'expected "application/json"'
+                        )
+
+                    logger.debug("Got metadata for uri=%s", uri_http)
+
+                    # Some requests don't provide `Content-Length` on the headers
+                    if len(response.content) > self.METADATA_MAX_CONTENT_LENGTH:
+                        raise MetadataRetrievalException(
+                            f"Retrieved content for uri={uri_http} is too big"
+                        )
+
+                    return response.json()
+            except (IOError, ValueError) as e:
+                django_cache.set(
+                    self.get_redis_metadata_timeout_key(uri),
+                    "Metadata timeout",
+                    60 * 60 * 24 * 2,
+                )
+                chance_to_retrieve_metadata_from_uri.apply_async(
```

Comment:
The idea here is to give more opportunities when the error is a timeout, currently configured at 15 seconds.